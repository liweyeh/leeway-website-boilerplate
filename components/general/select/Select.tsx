/**
 * select component
 * @author Liwei Yeh <liwei.yeh@adasasistemas.com>
 */

// Library
import { useState, useEffect } from 'react';

// CSS Style for this module
import styles from './select.module.scss';

interface SelectProps {
	/** Custom styling  */
	customStyles?: string;
	/** Label of the select list  */
	label?: string;
	/** Select options of the select list  */
	options?: string[];
	/** Is there an error  */
	error?: boolean;
	/** Default text that is displayed within the field */
	placeholder?: string;
	/** Msg to be displayed when the status is error */
	errorMsg?: string;
	/**
	 * onChange function for keeping track field value
	 * debounce is built in
	 */
	handleChange?: (value: string) => void;
}

// TODO: replace div tag with semantic HTML
/**
 * A dropdown list
 * @param {SelectProps} Prop
 * @returns  A dropdown list
 */
export const Select = ({
	customStyles,
	label = '',
	options = [],
	error = false,
	placeholder = 'Please select a value',
	errorMsg = 'This field cannot be empty',
	handleChange = () => {
		return;
	},
}: SelectProps) => {
	const [opened, setOpened] = useState<boolean>(false);
	const [selected, setSelected] = useState<string>();
	const [clicked, setClicked] = useState<boolean>(false);

	useEffect(() => {
		const closeSelect = () => {
			setOpened(false);
		};
		document.addEventListener('click', closeSelect);

		return () => {
			document.removeEventListener('click', closeSelect);
		};
	}, []);
	return (
		<div className={[styles.select, customStyles].join(' ')}>
			<div className={styles[`select-field`]}>
				{label && (
					<label className={styles[`select-name`]}>
						<p>{label}</p>
					</label>
				)}
				<div
					className={[
						styles[`select-container`],
						!clicked && styles[`select-container-default`],
						opened && styles[`select-container-opened`],
						error && styles['select-error'],
					].join(' ')}
					role={'listbox'}
					onClick={e => {
						e.stopPropagation();
						setOpened(prev => !prev);
					}}
				>
					<div className={styles[`select-container-text`]}>{selected ? selected : placeholder}</div>
				</div>
				{opened && (
					<div className={styles['select-list']}>
						{options.map((option, index) => (
							<div
								className={[
									styles['select-list-item'],
									option === selected && styles['select-list-item-selected'],
								].join(' ')}
								onClick={() => {
									handleChange(option);
									setSelected(option);
									setClicked(true);
								}}
								role="option"
								aria-selected={option === selected}
								key={index}
							>
								{option}
							</div>
						))}
					</div>
				)}
			</div>
			<div
				className={[styles['select-error-text'], !error && styles['select-error-text-hide']].join(
					' ',
				)}
			>
				{errorMsg}
			</div>
		</div>
	);
};
