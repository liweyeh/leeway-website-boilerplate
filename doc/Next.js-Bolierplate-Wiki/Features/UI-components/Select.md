[[_TOC_]]

# Overview

- Purpose: A dropdown select which is commonly used in form
- Type: General UI
- Status: ðŸš€ **Completed**
- Code: [URL](https://dev.azure.com/ADASA-Accelerator/Website-React-Booster/_git/website-boilerplate-nextjs?path=/components/general/select)
- Created by: Liwei Yeh <liwei.yeh@adasasistemas.com>
- Created time: 25/08/2022

# Props

- customStyles<`string`>: Custom styling
- label<`string`>: Label of the select list
- options<`string[]`>: Select options of the select list
- error<`boolean`>: Is there an error
- placeholder<`string`>: Default text that is displayed within the field
- errorMsg<`string`>: Msg to be displayed when the status is error
- handleChange<`(value: string) => void`>: onChange function for keeping track field value; debounce is built in

# Usage

```jsx
<Select
	customStyles={styles.select}
	label={'Error Code:'}
	options={['404', '500']}
	handleChange={v => {
		setStatus(v);
	}}
/>
```
