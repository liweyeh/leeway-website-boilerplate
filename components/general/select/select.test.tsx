import { render } from '@testing-library/react';
import { Select } from './Select';
import '@testing-library/jest-dom';

const setup = (error: boolean) => {
	const { asFragment } = render(
		<Select
			label="Select faulty meter."
			options={['a', 'b', 'c']}
			placeholder="Select"
			errorMsg="Please select an item"
			error={error}
			handleChange={() => {
				return;
			}}
		/>,
	);
	return { asFragment };
};

test('test Select snapshot', () => {
	const { asFragment } = setup(false);
	expect(asFragment()).toMatchSnapshot();
});

test('test error Select snapshot', () => {
	const { asFragment } = setup(true);
	expect(asFragment()).toMatchSnapshot();
});
