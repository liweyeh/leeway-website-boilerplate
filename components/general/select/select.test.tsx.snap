// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Select snapshot 1`] = `
<DocumentFragment>
  <div
    class="select "
  >
    <div
      class="select-field"
    >
      <label
        class="select-name"
      >
        <p>
          Select faulty meter.
        </p>
      </label>
      <div
        class="select-container select-container-default false false"
        role="listbox"
      >
        <div
          class="select-container-text"
        >
          Select
        </div>
      </div>
    </div>
    <div
      class="select-error-text select-error-text-hide"
    >
      Please select an item
    </div>
  </div>
</DocumentFragment>
`;

exports[`test error Select snapshot 1`] = `
<DocumentFragment>
  <div
    class="select "
  >
    <div
      class="select-field"
    >
      <label
        class="select-name"
      >
        <p>
          Select faulty meter.
        </p>
      </label>
      <div
        class="select-container select-container-default false select-error"
        role="listbox"
      >
        <div
          class="select-container-text"
        >
          Select
        </div>
      </div>
    </div>
    <div
      class="select-error-text false"
    >
      Please select an item
    </div>
  </div>
</DocumentFragment>
`;
